src/server/game/Movement/MovementGenerators/TargetedMovementGenerator.cpp

```diff
@@ -78,166 +92,162 @@ bool ChaseMovementGenerator<T>::DoUpdate(T* owner, uint32 time_diff)
+            float additionalRange = 0;
+            bool predictDestination = !mutualChase && target->isMoving();
+            if (predictDestination)
             {
-                cOwner2->SetCannotReachTarget(this->i_target.getTarget()->GetGUID());
+                UnitMoveType moveType = MOVE_RUN;
+                if (target->CanFly())
+                    moveType = target->HasUnitMovementFlag(MOVEMENTFLAG_BACKWARD) ? MOVE_FLIGHT_BACK : MOVE_FLIGHT;
+                else
+                {
+                    if (target->IsWalking())
+                        moveType = MOVE_WALK;
+                    else
+                        moveType = target->HasUnitMovementFlag(MOVEMENTFLAG_BACKWARD) ? MOVE_RUN_BACK : MOVE_RUN;
+                }
+                float speed = target->GetSpeed(moveType) * 0.5f;
+                additionalRange = owner->GetExactDistSq(target) < G3D::square(speed) ? 0 : speed;
             }
-        }
-
-        i_recalculateTravel = false;
-        i_path = nullptr;
-    }
-
-    if (_lastTargetPosition && i_target->GetPosition() == _lastTargetPosition.value() && mutualChase == _mutualChase)
-        return true;
-
-    _lastTargetPosition = i_target->GetPosition();
-
-    if (PositionOkay(owner, target, maxRange, angle) && !owner->HasUnitState(UNIT_STATE_CHASE_MOVE))
-    {
-        if (Creature* cOwner2 = owner->ToCreature())
-        {
-            cOwner2->SetCannotReachTarget();
-        }
-
-        return true;
-    }
-
-    float tarX, tarY, tarZ;
-    target->GetPosition(tarX, tarY, tarZ);
-
-    bool withinRange = owner->IsInDist(target, maxRange);
-    bool withinLOS   = owner->IsWithinLOS(tarX, tarY, tarZ);
-    bool moveToward  = !(withinRange && withinLOS);
 
-    _mutualChase = mutualChase;
-
-    if (owner->HasUnitState(UNIT_STATE_CHASE_MOVE))
-    {
-        // can we get to the target?
-        if (cOwner && !target->isInAccessiblePlaceFor(cOwner))
-        {
-            cOwner->SetCannotReachTarget(target->GetGUID());
-            cOwner->StopMoving();
-            i_path = nullptr;
-            return true;
-        }
-    }
+            float x, y, z;
+            bool shortenPath;
 
-    if (!i_path || moveToward != _movingTowards)
-        i_path = std::make_unique<PathGenerator>(owner);
-    else
-        i_path->Clear();
+            // if we want to move toward the target and there's no fixed angle...
+            if (moveToward && !angle)
+            {
+                // ...we'll pathfind to the center, then shorten the path
+                target->GetPosition(x, y, z);
+                shortenPath = true;
+            }
+            else
+            {
+                // otherwise, we fall back to nearpoint finding
+                target->GetNearPoint(owner, x, y, z, (moveToward ? maxTarget : minTarget) - chaseRange - additionalRange, 0, angle ? target->ToAbsoluteAngle(angle->RelativeAngle) : target->GetAngle(owner));
```

Comment:
Actually, the main mistake is here.
You can give it a try.
This is how I did it:
```
    /*bool shortenPath;
    // if we want to move toward the target and there's no fixed angle...
    if (moveToward && !angle)
    {
        // ...we'll pathfind to the center, then shorten the path
        target->GetPosition(x, y, z);
        shortenPath = true;
    }
    else
    {
        // otherwise, we fall back to nearpoint finding
        target->GetNearPoint(owner, x, y, z, (moveToward ? maxTarget : minTarget) - hitboxSum, 0, angle ? target->ToAbsoluteAngle(angle->RelativeAngle) : target->GetAngle(owner));
        shortenPath = false;
    }*/
    bool shortenPath;
    if (moveToward && !angle)
    {
        target->GetRandomContactPoint(owner, x, y, z);
        shortenPath = true;
    }
    else
    {
        target->GetNearPoint(owner, x, y, z, (moveToward ? maxTarget : minTarget) - hitboxSum, 0.0f,  angle ? target->ToAbsoluteAngle(angle->RelativeAngle) : target->GetAngle(owner));
        shortenPath = false;
    }
```